apps/rush-lib/src/cli/utilities/pnpm/PnpmLinkManager.ts

```diff
@@ -0,0 +1,210 @@
+        if (!localLinks) {
+          localLinks = [];
+          rushLinkJson.localLinks[localPackage.name] = localLinks;
+        }
+        localLinks.push(dependencyName);
+
+        // We did not find a suitable match, so place a new local package that
+        // symlinks to the Rush project
+        const newLocalFolderPath: string = path.join(localPackage.folderPath, 'node_modules', dependencyName);
+
+        const newLocalPackage: BasePackage = BasePackage.createLinkedPackage(
+          dependencyName,
+          matchedVersion,
+          newLocalFolderPath
+        );
+
+        newLocalPackage.symlinkTargetFolderPath = matchedRushPackage.projectFolder;
+        localPackage.children.push(newLocalPackage);
+      } else {
+        // weird state or program bug
+        throw Error('Cannot find rush dependency in rush configuration');
+      }
+    });
+
+    // Iterate through all the regular dependencies
+    Object.keys(commonPackage.packageJson!.dependencies || {}).forEach((dependencyName: string) => {
+      const dependencyVersionRange: string = commonPackage.packageJson!.dependencies![dependencyName];
+
+      // We can't symlink to an Rush project, so instead we will symlink to a folder
+      // under the "Common" folder
+
+      // here we need to get the list of versions and find the newest version that
+      // satisfies the dependency range (which is what I assume PNPM does)
+
+      const storePackageVersionsPath: string = path.join(
+        pnpmStorePath,
+        dependencyName
+      );
+
+      const availableVersions: string[] = fsx.readdirSync(storePackageVersionsPath).filter((version) => {
+        return fsx.lstatSync(path.join(storePackageVersionsPath, version));
+      });
+      availableVersions.sort((v1: string, v2: string) => {
+        return semver.gt(v1, v2) ? 1 : -1;
+      });
+
+      let selectedVersion: string;
+      for (const version of availableVersions) {
+        if (semver.satisfies(version, dependencyVersionRange)) {
+          selectedVersion = version;
+          break;
+        }
+      }
+
+      const pnpmStorePackagePath: string = path.join(
+        pnpmStorePath,
+        dependencyName,
+        selectedVersion!,
+        RushConstants.nodeModulesFolderName,
+        dependencyName
+      );
+
+      if (fsx.lstatSync(pnpmStorePackagePath).isDirectory()) {
+        const newLocalFolderPath: string = path.join(
+           localPackage.folderPath, 'node_modules', dependencyName);
+
+        const newLocalPackage: BasePackage = BasePackage.createLinkedPackage(
+          dependencyName,
+          selectedVersion,
+          newLocalFolderPath
+        );
+
+        newLocalPackage.symlinkTargetFolderPath = pnpmStorePackagePath;
+        localPackage.addChild(newLocalPackage);
+
+      } else {
+        throw Error(`The dependency "${dependencyName}" needed by "${localPackage.name}"`
+          + ` was not found the common folder -- do you need to run "rush generate"?`);
```

Comment:
>the [](start = 28, length = 3)

did you handle cyclic dependcies?

Reply:
this should be handled by virtue of the rushDependency/dependency distinction

---
In reply to: [152451652](https://github.com/Microsoft/web-build-tools/pull/449#discussion_r152451652) [](ancestors = 152451652)